### ISSUE TITLE
support graceful training cleanup after KeyboardInterrupt

### ISSUE THREAD

**Issue by jeremyjordan** [PyTorchLightning/pytorch-lightning#795]
## 🚀 Feature
In the training loop, we should catch `KeyboardInterrupt` exceptions and enter a graceful termination (ie. run `on_training_end`, print profiler report, etc). 

### Motivation

Sometimes I'll train a model and just let it run until something interesting happens or my loss has converged. Exiting via "Ctrl + C" can be easier than waiting for "EarlyStopping" to trigger in some cases. However, I may still want certain things to occur (eg. printing out the profiler summary) after stopping training. 

Salesforce's AWD-LSTM has an example of this feature implemented [here](https://github.com/salesforce/awd-lstm-lm/blob/master/main.py#L287). 

### Pitch

Figure out what actions we want to run after a KeyboardInterrupt. Then adjust the training loop to catch this exception and ensure the aforementioned actions still run afterwards.

**Comment by williamFalcon** [comment]
awesome, let's do it

**Comment by williamFalcon** [comment]
@jeremyjordan can we get it into this release? :)

**Comment by jeremyjordan** [comment]
when’s the release? planning to get to it this weekend

**Comment by williamFalcon** [comment]
feb 21

**Comment by jeremyjordan** [comment]
sure, shouldn't be a problem then :)

**Comment by lezwon** [comment]
Hey guys, I am currently training a model on Sagemaker and have added a notebook shutdown code within the `on_training_end` method. There were times where I had to manually cancel my model training because something was wrong. However, If I do that, the notebook shuts down. Is there any way to make this feature optional? I don't want my notebook shutting down after a keyboard interrupt, only after successful training completion.

**Comment by jeremyjordan** [comment]
@lezwon Hmm yeah, that's not ideal. Do you mind creating a new issue (you can reference this one) with the problem and I can draft up a PR to address it?
